From 157f245e96a7913250261a69de5b7bd6c9861673 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 22 Nov 2012 21:30:59 -0500
Subject: [PATCH] Misc EMC Changes

Only spawn 1 ocelot ever from eggs.
Load custom permissions before plugins
Unload plugins after everything else
Color name tags to match list
Make Durability enchant scale 0-80% reduction on Armor damage (up to level 5)
Zombies break down doors on normal
Update inventory on block place/interact cancellation
---
 src/main/java/net/minecraft/server/EntityOcelot.java          |  2 +-
 src/main/java/net/minecraft/server/ItemStack.java             |  9 +++++++++
 src/main/java/net/minecraft/server/MinecraftServer.java       | 11 ++++++-----
 .../net/minecraft/server/PacketPlayOutNamedEntitySpawn.java   |  4 +++-
 .../java/net/minecraft/server/PathfinderGoalBreakDoor.java    |  2 +-
 src/main/java/org/bukkit/craftbukkit/CraftServer.java         |  3 ++-
 .../java/org/bukkit/craftbukkit/event/CraftEventFactory.java  |  2 ++
 7 files changed, 24 insertions(+), 9 deletions(-)

diff --git a/src/main/java/net/minecraft/server/EntityOcelot.java b/src/main/java/net/minecraft/server/EntityOcelot.java
index 867abd7..3c3f28c 100644
--- a/src/main/java/net/minecraft/server/EntityOcelot.java
+++ b/src/main/java/net/minecraft/server/EntityOcelot.java
@@ -215,7 +215,7 @@ public class EntityOcelot extends EntityTameableAnimal {
 
     public GroupDataEntity a(GroupDataEntity groupdataentity) {
         groupdataentity = super.a(groupdataentity);
-        if (this.world.random.nextInt(7) == 0) {
+        if (false && this.world.random.nextInt(7) == 0) {
             for (int i = 0; i < 2; ++i) {
                 EntityOcelot entityocelot = new EntityOcelot(this.world);
 
diff --git a/src/main/java/net/minecraft/server/ItemStack.java b/src/main/java/net/minecraft/server/ItemStack.java
index 37ceade..b9d0db7 100644
--- a/src/main/java/net/minecraft/server/ItemStack.java
+++ b/src/main/java/net/minecraft/server/ItemStack.java
@@ -196,7 +196,16 @@ public final class ItemStack {
                 int j = EnchantmentManager.getEnchantmentLevel(Enchantment.DURABILITY.id, this);
                 int k = 0;
 
+                j = Math.min(5, j); // EMC
                 for (int l = 0; j > 0 && l < i; ++l) {
+                    // EMC start - make Unbreaking scale to level for armor. 5 = 80% chance of no loss.
+                    if (getItem() instanceof ItemArmor) {
+                        if (random.nextFloat() >= 1 - (.8 * (j / 5))) {
+                            k++;
+                        }
+                        continue;
+                    }
+                    // EMC end
                     if (EnchantmentDurability.a(this, j, random)) {
                         ++k;
                     }
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 022e032..ba49d8a 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -375,11 +375,6 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IMo
     public void stop() throws ExceptionWorldConflict { // CraftBukkit - added throws
         if (!this.L) {
             h.info("Stopping server");
-            // CraftBukkit start
-            if (this.server != null) {
-                this.server.disablePlugins();
-            }
-            // CraftBukkit end
 
             if (this.ag() != null) {
                 this.ag().b();
@@ -402,6 +397,12 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IMo
             }
             // CraftBukkit end */
 
+            // CraftBukkit start
+            if (this.server != null) {
+                this.server.disablePlugins();
+            }
+            // CraftBukkit end
+
             if (this.k.d()) {
                 this.k.e();
             }
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java b/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java
index 8bab528..1f6943c 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java
@@ -8,6 +8,7 @@ import java.io.IOException; // CraftBukkit
 
 public class PacketPlayOutNamedEntitySpawn extends Packet {
 
+    private EntityPlayer entityhuman = null; // EMC
     private int a;
     private GameProfile b;
     private int c;
@@ -22,6 +23,7 @@ public class PacketPlayOutNamedEntitySpawn extends Packet {
     public PacketPlayOutNamedEntitySpawn() {}
 
     public PacketPlayOutNamedEntitySpawn(EntityHuman entityhuman) {
+        this.entityhuman = (EntityPlayer) entityhuman; // EMC
         this.a = entityhuman.getId();
         this.b = entityhuman.getProfile();
         this.c = MathHelper.floor(entityhuman.locX * 32.0D);
@@ -50,7 +52,7 @@ public class PacketPlayOutNamedEntitySpawn extends Packet {
     public void b(PacketDataSerializer packetdataserializer) throws IOException { // CraftBukkit - added throws
         packetdataserializer.b(this.a);
         packetdataserializer.a(this.b.getId());
-        packetdataserializer.a(this.b.getName().length() > 16 ? this.b.getName().substring(0, 16) : this.b.getName()); // CraftBukkit - Limit name length to 16 characters
+        packetdataserializer.a(entityhuman.listName.length() > 16 ? entityhuman.listName.substring(0, 16) : entityhuman.listName); // EMC
         packetdataserializer.writeInt(this.c);
         packetdataserializer.writeInt(this.d);
         packetdataserializer.writeInt(this.e);
diff --git a/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java b/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java
index 60be62b..7a079f9 100644
--- a/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java
+++ b/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java
@@ -43,7 +43,7 @@ public class PathfinderGoalBreakDoor extends PathfinderGoalDoorInteract {
             this.j = i;
         }
 
-        if (this.i == 240 && this.a.world.difficulty == EnumDifficulty.HARD) {
+        if (this.i == 240 && this.a.world.difficulty.ordinal() >= EnumDifficulty.NORMAL.ordinal()) {
             // CraftBukkit start
             if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityBreakDoorEvent(this.a, this.b, this.c, this.d).isCancelled()) {
                 this.c();
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index b482592..59557ef 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -276,6 +276,7 @@ public final class CraftServer implements Server {
         if (type == PluginLoadOrder.STARTUP) {
             helpMap.clear();
             helpMap.initializeGeneralTopics();
+            loadCustomPermissions(); // EMC - Plugins should have access to this
         }
 
         Plugin[] plugins = pluginManager.getPlugins();
@@ -288,7 +289,7 @@ public final class CraftServer implements Server {
 
         if (type == PluginLoadOrder.POSTWORLD) {
             commandMap.registerServerAliases();
-            loadCustomPermissions();
+            // loadCustomPermissions(); // EMC - Moving up
             DefaultPermissions.registerCorePermissions();
             helpMap.initializeCommands();
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 161506a..bea89e4 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -117,6 +117,7 @@ public class CraftEventFactory {
         BlockPlaceEvent event = new BlockPlaceEvent(placedBlock, replacedBlockState, blockClicked, player.getItemInHand(), player, canBuild);
         craftServer.getPluginManager().callEvent(event);
 
+        if (event.isCancelled()) { player.updateInventory(); } // EMC
         return event;
     }
 
@@ -211,6 +212,7 @@ public class CraftEventFactory {
         PlayerInteractEvent event = new PlayerInteractEvent(player, action, itemInHand, blockClicked, blockFace);
         craftServer.getPluginManager().callEvent(event);
 
+        if (action == Action.RIGHT_CLICK_BLOCK && event.isCancelled()) { player.updateInventory(); } // EMC
         return event;
     }
 
-- 
1.8.3.2

